Add a native Rust Linux hidraw backend to replace hidapi

carlosmn opened the pull request:
I've been meaning to get around to this and just had a few free days so here we are. My ultimate goal is being able to have something on which we can do sensible async. At worst maybe we can add a `AsRawFd` but we can probably do better.

I've been away from the device I normally use so I haven't been able to test the IO portion itself, but I have checked the enumeration shows the same via the `lshid` example including the page and usage page values. The actual IO should be trivially correct as we're just asking the kernel to do the IO (though the report get situation with the buffer is a bit suspicious). I've tried to keep the behaviour the same as the C hidapi by using `nix` to capture the system calls more accurately instead of letting the Rust stdlib make it nicer for us.

I've added this as a new feature/backend in case there are particularities of this implementation that means you might want to go back to hidapi. It also made it easier to check both backends build and enumerate the same. The name is probably not the best, but I started by dealing with libudev so that's what I named it. It probably would make sense to rename it if/when this gets merged. Eventually hopefully this can be turned on by default, but I'm not sure how this project would want to deal with setting the defaults.

I see there were a couple of attempts at doing some generic-ish backend work and bypassing hidapi. I looked at a branch or two, but ended up building the interaction in a way that felt like it made sense to me. Hopefully it's sensible for others too.

The `Box<dyn ...>` does mean an extra allocation but I think the alternative of `struct HidDevice<T: HidDeviceBackend>` would break API. An alternative would also be to embed the backend `HidDevice` and change the type of the field depending on the backend at compile time. That wouldn't really use the trait to make sure you've implemented everything though, but that's not the end of the world as it's all internal types anyway (and there are workarounds you can do in tests to enforce that, I think).

I added a couple of test files from a wireless mouse that I happen to have nearby to test that the usage and usage_page values are the same as hidapi.

carlosmn committed:
Add a linux-shared-udev feature

This is to enable the version that bypasses the C hidapi and talks to udev
directly. Initially it does not bypass anything but it will grow features as it
can.

carlosmn committed:
Add enumeration via libudev

This is the first step in bypassing the hidraw library. It does not calculate
the `usage` or `page_usage` fields.

carlosmn committed:
Add a skeleton for a HidDevice bypassing the C hidapi

It's all `todo!()` but it's enough methods that the examples compile cleanly.

carlosmn committed:
Teach rust linux backend to open a device

carlosmn committed:
Add read and write support in rust linux backend

carlosmn committed:
Use nix when using the rust hidapi backend

carlosmn committed:
Add feature report get and set to rust hidraw backend

carlosmn committed:
Add read_timeout to the rust hidapi backend

carlosmn committed:
Fill in the information methods in the rust hidapi backend

carlosmn committed:
Support setting blocking or non-blocking in rust hidapi backend

We always open in non-blocking mode and build the blocking read on top of that.

carlosmn committed:
Add some From impls to HidError to handle the rust errors

carlosmn committed:
Make all methods of rust hidapi backend take `&self`

This lets the examples compile. The use of the `RefCell` might also be avoided
later if we go use the syscalls instead of going through Rust's io methods.

carlosmn committed:
Add a trait for backends to implement

This is just a copy of the public API we have right now. Later we might refine
it. We wrap the inner type dynamically (instead of e.g. using generics) so we
don't change the kind of struct that is shown publicly.

This also takes us back to a situation where the public API is documented in one
place rather than in the backend.

carlosmn committed:
Move the hidadpi backend to its own module

This lets us avoid the special-casing in `lib.rs` and we can export whichever
type exists for our backend.

This is hoever not great as the iteration still knows a bit too much, plus we
have no way to statically verify that the types have the same API.

carlosmn committed:
No need to link against hidadpi-hidraw on the rust backend

carlosmn committed:
Store the last error in the rust backend HidDevice

This mimics what we get from the C hidap because we expose this method in the
library.

carlosmn committed:
Move the HidApi implemenatation into each backend

Our use of HidApiBackend works a bit like a trait by failing at compile time if
a bckend has not implemented it. But it's not as clear-cut as having a specific
list of functions in a block.

carlosmn committed:
Use more raw syscalls in the rust hidapi backend

Try to be more like the C hidapi and use low-level readn and write so we get
access to errno to match the behaviour.

It also lets us avoid the `RefCell<File>` by just having the file descriptor.

carlosmn committed:
Perform the HIDRAW ioctl check in the rust backend

This is what the C hidapi uses to check that the device responds to the
appropriate messages.

carlosmn committed:
Quieten dead code warning in linuxudev

carlosmn committed:
Add a rust implementation for parsing report descriptors

This is a pretty straight port of the C code which can probably be made a bit
more rusty later.

carlosmn committed:
Get the first report descriptor in the linux rust backend

This just gets the first one. The C library then goes in and creates a new
`DeviceInfo` for every other usage it can find, but we will do that in a later
commit.

carlosmn committed:
Return multiple DeviceInfos for multi-usage in the rust linux backend

This matches the behaviour of the C library.

carlosmn committed:
Use the built-in little-endian conversion

This is what the code we ported from C was doing but we can be more explicit
here.

carlosmn committed:
Make the HID descriptor parsing a bit more rusty

Instead of passing in the position, we can pass in the rest of the slice.

carlosmn committed:
Add a test for HidrawReportDescriptor

Since this code is now parsing inside Rust, we can't punt the testing to the
external library.

Copy in the report descriptors for a wireless mouse I happen to have around and
compare against the numbers that we generate right now with both the C and Rust
implementation.

carlosmn committed:
Avoid copying the serial number when opening a device

ruabmbua commented on src/linux_udev.rs:
```diff
@@ -0,0 +1,762 @@
+use super::{BusType, DeviceInfo, HidDeviceBackend, HidError, HidResult, WcharString};
+
+/// Global error to simulate what C hidapi does
+static GLOBAL_ERROR: Mutex<RefCell<Option<String>>> = Mutex::new(RefCell::new(None));
```
You really do not need to emulate the C API, it would probably be a better idea to fix the API and make a breaking change. Why make the effort of implementing this in rust, and then keep bad C API practices ;-)

I imagine it would be pretty simple to remove check_error() API's when the udev backend is enabled.  For any actual legitimate use of this API, better to return the errors directly in all functions via Result.

carlosmn replied:
I agree, I was wondering if it might be too forward to propose a PR to deprecate it, since it doesn't seem to return anything that you wouldn't have in a better way.

We could deprecate it in mainline and make it a no-op or just not even have it with non-C-hidapi backends if and when they come in. Mostly I added it because one of the examples uses it, so it would actually compile and run. If it goes away selectively then probably that would have to change.

carlosmn replied:
Well it looks like this wouldn't be the first method that disappears depending on the backend, so maybe it's fine to just not expose it.

ruabmbua commented on src/linux_udev.rs:
```diff
@@ -0,0 +1,762 @@
+//! Functions for talking to udev directly instead of going through hidapi
```
Probably not a good name for the backend.

I would just call it linux.rs, or maybe linux_native.rs?

ruabmbua commented on src/linux_udev.rs:
```diff
@@ -0,0 +1,762 @@
+}
+
+// Bus values from linux/input.h
+const BUS_USB: u16 = 0x03;
```
But consts at the top of the file.

ruabmbua commented on src/linux_udev.rs:
```diff
@@ -0,0 +1,762 @@
+}
+
+// From linux/hidraw.h
+const HIDRAW_IOC_MAGIC: u8 = b'H';
```
put consts at top of file.

Even better, maybe put these ioctl defs and related consts into a external file.

carlosmn replied:
Sure, but nothing other than linux-native would use ioctls, no?

ruabmbua replied:
yep these are pretty linux specific

carlosmn replied:
Moved these to their own file that we then use in linux_native.

ruabmbua commented on src/linux_udev.rs:
```diff
@@ -0,0 +1,762 @@
+/// These strings would be of the format
+///     type vendor   product
+///     0003:000005AC:00008242
+fn parse_hid_vid_pid(s: &str) -> Option<(u16, u16, u16)> {
```
Probably a good idea to out parsing related functionality (which might even be shared between backends) into a different file.
Also the unit tests for parsing could be moved to that file.

carlosmn replied:
I've left the things here because I don't know if a macos_native backend would even need this or if the system libraries would parse a lot more for us.

carlosmn replied:
Looking at what C hidapi does for macOS, it doesn't look like we'd need them for anything else. The vendor and product ids are parsed by the system and given to us, as are the usage_page and usage information. Same goes with Windows.

So we can move them somewhere else if you prefer, like with the ioctls, but it doesn't look like it's something we will be re-using.

ruabmbua commented on Cargo.toml:
```diff
@@ -23,13 +23,18 @@ linux-static-libusb = []
 linux-static-hidraw = []
 linux-shared-libusb = []
 linux-shared-hidraw = []
+linux-shared-udev = ["dep:udev", "dep:nix"]
```
hm, probably not a good name for the backend. Maybe rename it to just linux, and make it clear in the lib.rs documentation that the *linux* backend is a native rust implementation (mostly) without C dependencies.

ruabmbua reviewed (commented):
Just browsed through it at a quick glance, and found a few things (mostly code structure).
I will do a more in depth review when I have more time.

But thank you very much, getting C dependencies out of the way and then refactoring the API a bit would make the library much simpler to use (less dependency issues with multiple hidapi versions in the dep tree).

ruabmbua commented on src/lib.rs:
```diff
@@ -41,6 +41,7 @@
 //! - `linux-static-hidraw`: uses statically linked `hidraw` backend on Linux (default)
 //! - `linux-shared-libusb`: uses dynamically linked `libusb` backend on Linux
 //! - `linux-shared-hidraw`: uses dynamically linked `hidraw` backend on Linux
+//! - `linux-shared-udev`: uses dynamically linked `udev` backend on Linux
```
udev is not really a backend for this, I`d rather call the backend *linux* and describe it as having no dependencies to the hidapi C library.

carlosmn committed:
Rename the native Linux backend to 'linux_native'

The udev naming was misleading and just calling it linux_native makes it much more
accurate as to how we're talking to the devices. This also opens the naming for
other systems with a native and non-native version.

carlosmn commented:
I've renamed the backend and options to simply `linux` as ideally that would be the only/default one there.

carlosmn committed:
Get rid of check_error when not using C hidapi

In the native Linux implementation there is no need to keep doing this behaviour
from C hidapi so get rid of this and let's not even expose it in the public API.

carlosmn commented:
It looks like naming an option `linux` makes clippy think we're confused about what we mean, so now it's named `linux-native` and `linux_native`. This is probably for the best as it makes space in the naming for things like `macos_native`.

I've also edited some earlier commits so we open the right file so things do work.

Unfortunately by having the path be the dev node, i.e. `/dev/hidrawN`, the udev library doesn't seem to have a way to open based off of that, as it only has `Device::from_syspath` as a way to create the device directly, so we don't actually have enough information, so `get_device_info()` is currently simply broken.

carlosmn commented:
Oh I also made it so the `check_error` methods are only available when using C hidapi.

ruabmbua commented:
Regarding udev problem with the devnode, can you maybe elaborate more? I think I had no issues with that when I was working on this myself. It is also poissible to map from /dev file to /sys path with udev (although I do not remember the exact API for that).

carlosmn commented:
In order to match the C hidapi does, and so you can pass the `DeviceInfo::path` into `HidApi::open_path` and have it work then we need to open via the dev node, and that's what we can store in our struct. I don't know off hand if the files under `/sys` can be opened to use as devices.

The way C hidapi then can get the info is via `udev_device_new_from_devnum` and then filling is as usual, but that constructor is not available in the Rust udev library. I don't see a way to go from one to the other without going through and listing them all, but I'm not that familiar with the layouts here.

carlosmn commented:
It looks like it should be easy yeah https://github.com/systemd/systemd/blob/5015b5014bcff93371aef2c78b92efcfc2e38a40/src/libsystemd/sd-device/sd-device.c#L304 though it would be better if the udev library supported it.

carlosmn committed:
Remove useless extern declaration

carlosmn committed:
Open the device via major/minor in get_device_info

This is what happens internally in libudev but that constructor is not currently
accessible form Rust so we do it ourselves.

carlosmn committed:
Move ioctls to their own file

They only get loaded for linux-native as that's the only thing that can use
them.

carlosmn committed:
Move consts to the beginning of the file

carlosmn committed:
Don't gate the linux-native dependencies on an arch

carlosmn commented:
It turns out this completely broke macOS (and probably Windows) because they want to add methods so I guess I'll see if I can tackle that.

carlosmn committed:
Use hidapi to load C hidapi rather than not(linux_native)

carlosmn committed:
Let macOS have its own trait with extensions for HidDevice

While moving things into backends for linux-native, we never addressed the needs
of macOS or Windows where they have extra functionality.

Let us build on macOS again by adding specific traits for it.

carlosmn committed:
Let windows have its own trait with extensions for HidDevice

Like with the macOS-specific functions, here's the extras for Windows.

carlosmn commented:
It's not super clean but I've made it so each os can implements its own extras.

ruabmbua commented on build.rs:
```diff
@@ -77,19 +78,29 @@ fn compile_linux() {
+            "LINUX_NATIVE",
+            Box::new(|| {
+                // The udev crate takes care of finding its library
+                println!("cargo:rustc-cfg=linux_native");
```
Since linux_native is already a feature, this is not necessary. You can use the usual cfg(feature) attribute

carlosmn replied:
It's following the pattern elsewhere of adding the extra cfg but yeah I think it makes sense to skip it.

ruabmbua commented on src/error.rs:
```diff
@@ -50,3 +50,20 @@ impl Display for HidError {
+
+impl From<std::io::Error> for HidError {
+    fn from(e: std::io::Error) -> Self {
+        Self::HidApiError {
```
You can a new io error enum variant, and use that instead of turning the error into a string. Allows the user to work with errors (e.g. for async extract EWOULDBLOCK / EAGAIN etc...).

carlosmn replied:
I didn't want to add to the public API here and break error handling that users of the library might already have, but for this it's probably OK since making use of more rust probably means this is going to be a fairly normal error that we'll have to deal with.

carlosmn replied:
I've added a variant for this as it's generally useful and we'll need it more as we write more Rust.

ruabmbua replied:
Ok thx

ruabmbua commented on src/error.rs:
```diff
@@ -50,3 +50,20 @@ impl Display for HidError {
+
+#[cfg(linux_native)]
+impl From<nix::errno::Errno> for HidError {
+    fn from(e: nix::errno::Errno) -> Self {
```
Probably also a better idea to properly store the error in the enum

carlosmn replied:
This feels like a different situation from `std::io::Error` as it would add a variant that only exists with some options, which seems like it's going to make it a pain for users of the library to deal with. But if `std::io::Error` does become a variant, then maybe we can make a decent effort at translating it to that.

carlosmn replied:
Given that any nix Errno is going to be convertible into a `std::io::Errror` and they provide a impl for that, I've made it so nix errors convert into the `IoError` variant which lets us avoid an extra equivalent variant.

ruabmbua replied:
Good idea, lets do it like that

ruabmbua commented on src/lib.rs:
```diff
@@ -57,21 +58,32 @@
+mod hidapi;
+#[cfg(linux_native)]
+mod ioctl;
+#[cfg(linux_native)]
```
probably better to not use manual cfg when not necessary. Can use cfg(feature)

ruabmbua commented on src/lib.rs:
```diff
@@ -287,74 +250,16 @@ impl HidApi {
+#[derive(Clone, PartialEq)]
 enum WcharString {
     String(String),
+    #[cfg_attr(linux_native, allow(dead_code))]
```
You could also just remove this enum variant in case of linux_native?

carlosmn replied:
I was trying to avoid changing API but I now realise this is just internal to the library so that's maybe a slightly more efficient option.

carlosmn replied:
Ok I just tried this, but if we remove `Raw` on linux-native, then all the other code that wants to use it in `DeviceInfo` breaks and that is public API so it seems better to leave it at allowing dead code.

ruabmbua commented on src/lib.rs:
```diff
@@ -514,55 +419,85 @@ impl fmt::Debug for DeviceInfo {
+#[cfg(not(any(target_os = "macos", target_os = "windows")))]
+impl<T> HidDeviceBackend for T where T: HidDeviceBackendBase {}
+
+/// Automatically implement the top trait
```
either only use this comment one time, or all three times

carlosmn replied:
Ah, this came from doing it in a VM where copy-paste was annoying so I missed out on the comment.

ruabmbua commented on src/linux_native.rs:
```diff
@@ -0,0 +1,670 @@
+
+impl HidApiBackend {
+    pub fn get_hid_device_info_vector() -> HidResult<Vec<DeviceInfo>> {
+        // The C version assumes these can't fail, and they should only fail in case
```
currently udev might only fail on memory alloc failure (which is hard to believe?).
You can also forward any errors happening in udev to the Result, and embed the udev error inside the enum.

carlosmn replied:
Right, on Linux you would basically never see an actual malloc error, so we could just blow up there. I'm not a fan of adding udev error structs into the library's errors for the same reason as nix errors, where they exist just on the one OS.

ruabmbua replied:
Ok sure its ok

ruabmbua commented on src/linux_native.rs:
```diff
@@ -0,0 +1,670 @@
+            Err(_) => return Ok(Vec::new()),
+        };
+
+        let mut devices = Vec::new();
```
Maybe prefer a .filter_map() + collect for this code.
In this case the iterator code can give the vec hints on how much to reserve. Iterator code is also more idomatic.

carlosmn replied:
We return a vec so we still need a flatten after this. I suppose we could return an empty vector instead of a None and then a `flatten_map` by itself would work, but the semantics are slightly different then.

ruabmbua commented on src/linux_native.rs:
```diff
@@ -0,0 +1,670 @@
+///     type vendor   product
+///     0003:000005AC:00008242
+fn parse_hid_vid_pid(s: &str) -> Option<(u16, u16, u16)> {
+    let elems: Vec<Result<u16, _>> = s.split(':').map(|s| u16::from_str_radix(s, 16)).collect();
```
Do not create  a vector from the iterator. Instead store the iterator in a variable  and call .next() three times filling up the tuple.

carlosmn replied:
With all the enums it gets awkward but I made a macro to behave like let-else so it's a bit nicer to read and we do avoid allocating the two vectors.

ruabmbua commented on src/linux_native.rs:
```diff
@@ -0,0 +1,670 @@
+    let path = match raw_device
+        .devnode()
+        .map(|p| p.as_os_str().to_os_string().into_vec())
+        .map(CString::new)
```
I am not too happy that the lib is using CStrings for paths. Should be OsString or rather PathBufs.

Unfortunately getting rid of that will not be that easy. Something to consider when  the library is fully converted to native backends.

carlosmn replied:
It would require a version bump due to breaking the API so yeah i think it makes sense to only do it once everything can change over.

ruabmbua replied:
yes lets keep it as is

ruabmbua commented on src/linux_native.rs:
```diff
@@ -0,0 +1,670 @@
+    // USB has a bunch more information but everything else gets the same empty
+    // manufacturer and the product we read from the property above.
+    let info = match bus_type {
+        BusType::Usb => fill_in_usb(raw_device, info, name),
```
Probably clearer to pass the info as &mut reference and not use a input + return parameter?

carlosmn replied:
It seemed more rusty for the function to take ownership and return the changed one, kinda like you do with a builder pattern. But updating the fields also works.

ruabmbua commented on src/linux_native.rs:
```diff
@@ -0,0 +1,670 @@
+}
+
+// From linux/hidraw.h
+struct HidrawReportDescriptor {
```
Why not simply store this in a vector or boxed slice?
You can still newtype HidrawReportDescriptor around it.

ruabmbua replied:
Also using u32 for a length in rust is weird. please use usize

carlosmn replied:
It's just what the C version uses, which is not what we would do fresh in Rust. We probably could thinly-wrap a `Vec<u8>` or something.

carlosmn replied:
This is now just wrapping a vector, so you could have descriptors larger than the kernel allows, but that's probably fine since the kernel is giving us the descriptor.

ruabmbua commented on src/linux_native.rs:
```diff
@@ -0,0 +1,670 @@
+            Ok(v) => v,
+            Err(_) => {
+                return Err(HidError::HidApiError {
+                    message: "HID report descriptor over 4kB".into(),
```
Not sure why there is a 4kb error message here. The try_into::<u32>() will be able  to convert much larger values.

ruabmbua replied:
yep this panics: 

```rust
    #[test]
    fn overflow() {
        let data = vec![0u8; 10 * 4096];
        let desc = HidrawReportDescriptor::from_slice(&data).expect("bla");
    }
}
```

carlosmn replied:
4kB is the max size of a HID report descriptor from sysfs, at last according to C hidapi. If something makes us read a file that's bigger, that's  probably a bug somewhere. That said, I think I missed out on an `if ` in the match or something because that error message definitely doesn't match the actual error.

carlosmn replied:
This went away with the other changes around the cursor.

ruabmbua commented on src/linux_native.rs:
```diff
@@ -0,0 +1,670 @@
+
+        self.usage_page = usage_page;
+        self.initial = false;
+        self.value = &self.value[advanced..];
```
probably makes more sense to store a offset instead of initial bool, and not resize the slice in the iterator, but rather track with offset?

Then initial property can be found by checking if offset == 0

ruabmbua replied:
Also this whole thing looks like it could use a `Cursor`, which would keep track of bytes read. Works like a file with .read() .seek() etc...

carlosmn replied:
I've ported it to use `Cursor` which is nicer ins some ways but in a bunch of places we really do just want to grab a single byte so we end up doing `cursor.bytes().next()` a couple of times, and the size to advance is also based off of the initial position, so a couple of places are a bit awkward, but it does remove a bit of state from the iterator itself.

ruabmbua commented on src/linux_native.rs:
```diff
@@ -0,0 +1,670 @@
+        return 0;
+    }
+
+    match num_bytes {
```
Probably makes more sense to not match but rather just do  this:

```rust
let mut bytes: [0u8; 4] = [0; 4];
bytes[..num_bytes].copy_from_slice(desc[..num_bytes]);
u32::from_le_bytes(bytes)
```

Also the num_bytes param may be unneeded? Why not just use the slice len

carlosmn replied:
I think that does work. It's clearer now that the endian conversion isn't done by hand. The `num_bytes` is the actual amount of data that we should read, while `desc` is the whole rest of the descriptor. Now, we should be able able to call this function with the equivalent of `&desc[..num_bytes]` in the parent, but I'm not sure where it's less awkward.

ruabmbua commented on src/linux_native.rs:
```diff
@@ -0,0 +1,670 @@
+
+    // Short Item. Bottom two bits contains the size code
+    match key & 0x03 {
+        v @ 0 | v @ 1 | v @ 2 => Some((v.into(), 1)),
```
v @ 0..=2 ?

ruabmbua commented on src/linux_native.rs:
```diff
@@ -0,0 +1,670 @@
+/// Get the attribute from the device and convert it into a i32
+///
+/// On error or if the attribute is not found, it returns -1;
+fn attribute_as_i32(dev: &udev::Device, attr: &str) -> i32 {
```
I would rather have API`s using Option / Result instead of negative error codes.

carlosmn replied:
Yeah, that'd make sense as a rustification pass.

ruabmbua commented on src/linux_native.rs:
```diff
@@ -0,0 +1,670 @@
+/// Get the attribute from the device and convert it into a u16
+///
+/// On error or if the attribute is not found, it returns 0;
+fn attribute_as_u16(dev: &udev::Device, attr: &str) -> u16 {
```
yep, please use options intead. Its hard to keep  track of all those error == 0, error < 0, error == -1 etc...
Not necessary with rust

ruabmbua commented on src/linux_native.rs:
```diff
@@ -0,0 +1,670 @@
+fn osstring_to_string(s: OsString) -> WcharString {
+    match s.into_string() {
+        Ok(s) => WcharString::String(s),
+        Err(_) => panic!("udev strings should always be utf8"),
```
are you sure? This is probably more related to the locale of the linux system, which does not force utf-8

carlosmn replied:
I can double-check but I think this is mostly about properties in the HID stuff that don't change (at least the ones we use).

carlosmn replied:
We do use this for the product and manufacturer, maybe those don't have to be utf-8 then. I wonder what to do here, since we don't have the wide chars for the `WcharString::Raw`. Should we do a lossy conversion? Convert to wchar if it's not utf-8?

ruabmbua commented on src/linux_native.rs:
```diff
@@ -0,0 +1,670 @@
+/// Object for accessing the HID device
+pub struct HidDevice {
+    blocking: Cell<bool>,
+    fd: OwnedFd,
```
Oh nice new struct from new rust, did not know about it

carlosmn replied:
The worry here would be to increase the minimum version of Rust too much as 1.63, where this was stabilised, is less than a year old.

ruabmbua commented on src/linux_native.rs:
```diff
@@ -0,0 +1,670 @@
+        let fd: OwnedFd = match OpenOptions::new()
+            .read(true)
+            .write(true)
+            .custom_flags(libc::O_CLOEXEC | libc::O_NONBLOCK)
```
I am not sure CLOEXEC is useful to pass as a default?

carlosmn replied:
It's a useful default in general. Without it, a child process would also receive a handle to any hid devices the parent process has open, which seems like a bad idea.

ruabmbua replied:
You are right, probably the better default. I use it everywhere in c++ also

ruabmbua commented on src/ioctl.rs:
```diff
@@ -0,0 +1,29 @@
+    hidraw_ioc_grdescsize,
+    HIDRAW_IOC_MAGIC,
+    HIDRAW_IOC_GRDESCSIZE,
+    i32
```
probably should use the c_int type instead to make sure this matches with the kernel on all weird archs?

carlosmn replied:
I don't know if there are arches where Linux's int isn't 32-bit but it doesn't hurt.

ruabmbua commented on src/linux_native.rs:
```diff
@@ -0,0 +1,670 @@
+            return Err(HidError::InvalidZeroSizeData);
+        }
+
+        Ok(write(self.fd.as_raw_fd(), data)?)
```
This might return EAGAIN / EWOULDBLOCK, even if the device should be running in blocking mode?

carlosmn replied:
Good question, this is what the C hidapi does so it might be fine? This isn't a regular file write but this ends up enqueuing into whatever queue the kernel has so I don't know if it this could block.

carlosmn replied:
Based on https://docs.kernel.org/hid/hidraw.html where it only mentions blocking on `read()` it sounds like this is more akin to putting stuff in a queue than dong IO.

ruabmbua replied:
Ok interesting. Although its always a bit weird when special devices in linux do not behave like e.g. the read() or write() manpages describe it.

But what happens if the queue is full? Maybe it will block then. Could be a theoretical possibility, that has no chance of practically happening. Lets keep it like it is, if we ever have weird issues in regards to writing, we can look again.

carlosmn replied:
Looking through a bit of kernel code, it looks like the `hidraw_write` function does a bunch of checking and you just end up doing the same as the send-report ioctl, and then it gets handed over to the appropriate driver to whatever callback it has decided. So it looks like what happens is whatever each device driver decides to do.

ruabmbua reviewed (changes requested):
I did  a more in depth review, hope its not too much ;-)

carlosmn commented:
Heh, it looks a bit too much for the PR view :) I'll go through the comments. A lot looks like the result of porting the C code to Rust and keeping some of its behaviour.

carlosmn committed:
Return Option in number-parsing attribute getters

These used to return a default value but we determine what that error value
should be in the caller now.

carlosmn committed:
Use cfg(feature) for linux-native instead of defining something new

carlosmn committed:
Use c_int in ioctl

carlosmn committed:
Add mssing comment on Windows

carlosmn committed:
Simply the HID report bytes function

All of it is doing the little-endian conversion so we let the types do this. It
does mean we don't enforce the valid numbers but those come from our own code
that only ever produces the correct ones.

carlosmn committed:
Make report descriptor parsing a bit more rusty

Thinly wrap a vector instead of copying the kernel type and use a cursor to
advance in it instead of doing manual array arithmetic.

carlosmn committed:
Make the device-to-info loop a bit more rusty

carlosmn committed:
Use less complex vector and iterator methods in parse_hid_vid_pid

Instead of taking two different vectors we can pattern-match on the result of
the iteration of the first one.

ruabmbua commented on src/linux_native.rs:
```diff
@@ -0,0 +1,640 @@
+fn hid_report_bytes(cursor: &mut Cursor<&Vec<u8>>, num_bytes: usize) -> u32 {
+    let mut bytes: [u8; 4] = [0; 4];
+    if cursor.read_exact(&mut bytes[..num_bytes]).is_err() {
+        return 0;
```
This should probably also be returned as an Err/option. And on the call site treat it as a parsing error discarding the whole item?

carlosmn replied:
Yes I suppose that's what you should be doing in Rust.

carlosmn committed:
Add a `std::io::Error` variant to Error

Currently the `Error` is mostly around C hidapi but as we move further into Rust
it will be increasingly useful to have something more expressive, or at least
something that's more straightforward to convert.

Add a variant with the std error and convert nix errors into it.

carlosmn committed:
Return an error instead of 0 in hid_report_bytes

ruabmbua commented on src/linux_native.rs:
```diff
@@ -0,0 +1,644 @@
+/// These strings would be of the format
+///     type vendor   product
+///     0003:000005AC:00008242
+fn parse_hid_vid_pid(s: &str) -> Option<(u16, u16, u16)> {
```
This also works:

```rust
fn parse_hid_vid_pid(s: &str) -> Option<(u16, u16, u16)> {
    let mut elems = s.split(':').map(|s| u16::from_str_radix(s, 16));
    let devtype = elems.next()?.ok()?;
    let vendor = elems.next()?.ok()?;
    let product = elems.next()?.ok()?;

    Some((devtype, vendor, product))
}
```

carlosmn replied:
oh huh, is this a newer feature? I thought I'd tried to do this at some point in the past but `Option` wasn't a thing that the question mark worked with.

ruabmbua replied:
Not sure when, but already a while I think.

ruabmbua commented:
Thanks for the work, I will merge this back now and do some small fixups. I will then release it to crates.io after I have some more time verifying it does not break anything (hopefully I have time soon).

carlosmn commented:
Awesome, glad we could get to something we're happy with.
